### PR TITLE
Support pytest for distribution testing

### PR DESCRIPTION
**Pull request by neerajprad**
In response to https://github.com/pytorch/pytorch/issues/11578. This is a test run to see if CI (and other internal systems) works fine with pytest style tests.
 - Creates a separate `distributions` directory within `test`.
 - For testing, this rewrites the `constraint` tests as parameterized tests in pytest. I don't plan to convert any other tests to pytest style, but only expose this option for adding new tests, if required.

If this is a success, we can move `EXAMPLES` in `test_distributions` into a separate file that can be imported by both pytest and unittest style tests. cc. @fritzo

### CODE REVIEW

File: test/run_test.py [hunk 1]
```diff
@@ -36,7 +37,8 @@
     'distributed/test_data_parallel',
     'distributed/test_distributed_fork',
     'distributed/test_distributed_spawn',
-    'test_distributions',
+    'distributions/test_constraints.py',
+    'distributions/test_distributions',
```
Comment (fritzo) [inline test/run_test.py:41]: is there a reason some files end with .py and others don't?

Comment (neerajprad) [reply test/run_test.py:41]: Sorry, that was a typo, will fix. The test should still run with or without .py.

**Comment by neerajprad** [comment]
@ezyang, @mruberry - This seems to work fine on CircleCI. Is there anything else that we need to check (not sure if this breaks anything internally)?

**Comment by ezyang** [comment]
PR looks like it's going to need some internal build system changes as locations of files have moved around

**Pull request closed without merging**